Build by vendoring libmagic

stanhu opened the pull request:
This commit downloads the `file` library from http://www.darwinsys.com/file/, builds and installs the files into
`lib/ext`, and links the C extension against it.

At the moment, this isn't staticallly linked because I ran into some linker issues on macOS, but it could be done later if desired. The `file` binary is present and could be removed as well.

Closes #1

stanhu committed:
Build by vendoring libmagic

This commit downloads the `file` library from
http://www.darwinsys.com/file/, builds and installs the files into
`lib/ext`, and links the C extension against it.

At the moment, this isn't staticallly linked because I ran into some
linker issues on macOS, but it could be done later if desired. The
`file` binary is present and could be removed as well.

kwilczynski commented:
Hi @stanhu, thank you for this!

This is very nice and a step in the right direction. The static linking issues is something I also run in the past - there is issues with linker, plus `mkmf` also stops being able to detect whether functions (symbols) are available, so the following check using `have_func()` would stop working correctly:

https://github.com/kwilczynski/ruby-magic/blob/6c110aaf93298e166475bbea4542d139ccf6916f/ext/magic/extconf.rb#L126-L136

But, I haven't looked into this since, apologies!

Krzysztof

stanhu commented:
Talk about a ton of builds! https://travis-ci.org/github/kwilczynski/ruby-magic/builds/764335565

I wonder why some of the builds are timing out; could it be Travis or the FTP site throttling connections?

stanhu commented:
Oh, I guess we can clean up `.travis-ci.yml` if we don't need to test every version.

stanhu committed:
Remove libmagic matrix testing in .travis.yml

Since we now vendor libmagic, we don't need to test every version known
to man.

kwilczynski commented:
Hi @stanhu,

> Talk about a ton of builds! [travis-ci.org/github/kwilczynski/ruby-magic/builds/764335565](https://travis-ci.org/github/kwilczynski/ruby-magic/builds/764335565)

Yes, sorry about this!

Since in the past I had to rely on supporting the version of `libmagic` that was present in the environment where `ruby-magic` could have been installed (and `libmagic` has some interesting issues in some of the older version) I was testing against versions that seemed to be most widely available.

> I wonder why some of the builds are timing out; could it be Travis or the FTP site throttling connections?

Could be both! I have odd issues with the free version of Travis CI in the past, but also the FTP server that hosts the `libmagic` has some limits I think (or is generally slow).

This is why I used a small build script that attempts to fetch `libmagic` source from multiple mirrors, and once it succeeds at least once, then next time it would leverage Travis CI build cache - which is not something that we have the benefit of when downloading source each time now.

Krzysztof

kwilczynski commented:
Hi @stanhu,

> Oh, I guess we can clean up `.travis-ci.yml` if we don't need to test every version.

Definitely. It's not really needed any more - perhaps we will revisit this in the future once moving to something like the `mini_portile` Ruby Gem, where people could pass a command-line switch to use libraries present already in the system (so we can test this scenario correctly too), but for the time being no issue with removing the extra versions and the test matrix.

I also need to think how to bring testing with both GCC and Clang back. :)

Krzysztof

stanhu commented:
We could still have a matrix test with different compilers. I think setting the `CC` variable would do that.

I think Nokogiri (https://github.com/sparklemotion/nokogiri/blob/main/ext/nokogiri/extconf.rb) may be a good example of allowing users to use the system libraries.

stanhu committed:
Bring back gcc and clang testing

stanhu committed:
Switch to HTTPS mirror

The previous FTP sites only could support 10 concurrent downloads.

kwilczynski commented:
Hi @stanhu,

> We could still have a matrix test with different compilers. I think setting the `CC` variable would do that.

We can do that and directly use `CC` - it should work - or keep it as I did it in the past.

> I think Nokogiri ([https://github.com/sparklemotion/nokogiri/blob/main/ext/nokogiri/extconf.rb](https://github.com/sparklemotion/nokogiri/blob/main/ext/nokogiri/extconf.rb?rgh-link-date=2021-03-25T06%3A19%3A06Z)) may be a good example of allowing users to use the system libraries.

I was looking at Nokogori as an example of how it's done - it seems few other projects adopted similar approach (some even based on Nokogiri's way of handling options, etc.), and I wanted (when I was initially thinking of vendoring `libmagic`) to keep things consistent.

Krzysztof

kwilczynski commented:
Hi @stanhu,

I uploaded `libmagic` source code to an S3 bucket to solve the timeout during download issues, as per:

- https://ruby-magic.s3.eu-central-1.amazonaws.com/file-5.39.tar.gz
- https://ruby-magic.s3.eu-central-1.amazonaws.com/file-5.39.tar.asc
- https://ruby-magic.s3.eu-central-1.amazonaws.com/file-5.39.tar.gz.SHA256SUM

But since I noticed you found a mirror. :)

Krzysztof

kwilczynski commented:
Hi @stanhu,

You should be able to merge at your convenience when you are happy with the changes. I added you as collaborator.

Thank you again for help!

Krzysztof

stanhu commented:
@kwilczynski Thanks! I'm playing around with `mini_portile2`, but those improvements can come later.

stanhu commented:
@kwilczynski Thanks for your help! Could you add me to RubyGems as well? https://rubygems.org/profiles/stanhu

kwilczynski commented:
Hi @stanhu,

> @kwilczynski Thanks for your help! Could you add me to RubyGems as well? [rubygems.org/profiles/stanhu](https://rubygems.org/profiles/stanhu)

Done! You should receive invite shortly.

Krzysztof

kwilczynski reviewed (commented):
Hi @stanhu, thank you for working on this!

jrochkind commented:
> I wonder why some of the builds are timing out; could it be Travis or the FTP site throttling connections?

Travis doesn't really offer free support to open source anymore, although they are having trouble being transparent about saying so, instead making it complicated to come to that determination. 

https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works

Github actions is one alternative.

olearycrew commented:
Another alternative is GitLab CI/CD 😏 

* https://about.gitlab.com/solutions/github/
* https://about.gitlab.com/solutions/open-source/

meinac commented on ext/magic/extconf.rb:
```diff
@@ -1,6 +1,42 @@
+
+unless File.exist?(filename)
+  File.open(filename, 'wb') do |target_file|
+    URI.open("https://fossies.org/linux/misc/file-#{LIBMAGIC_TAG}.tar.gz", 'rb') do |read_file|
```
@stanhu, @kwilczynski  - This line is failing with a 412 error because the `User-Agent` header seems like a mandatory header for FOSSIES.

I've found this by trying the following script on IRB;

```ruby
require 'httparty'

url = "https://fossies.org/linux/misc/file-5.39.tar.gz"
real_agent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36"
madeup_agent_1 = "Ruby"
madeup_agent_2 = "Ruby 2.7.0"

HTTParty.get(url).code # => 412
HTTParty.get(url, headers: { "User-Agent" => madeup_agent_1 }).code # => 412

HTTParty.get(url, headers: { "User-Agent" => madeup_agent_2 }).code # => 200
HTTParty.get(url, headers: { "User-Agent" => real_agent }).code # => 200
```

I think we should either include this tarball into the gem source or set the HTTP header as `RUBY_DESCRIPTION`
